rust/src/csvreader.rs

```diff
@@ -0,0 +1,243 @@
+        // read a batch of rows into memory
+        let mut rows: Vec<StringRecord> = Vec::with_capacity(self.batch_size);
+        for _ in 0..self.batch_size {
+            match self.record_iter.next() {
+                Some(Ok(r)) => {
+                    rows.push(r);
+                }
+                Some(Err(_)) => {
+                    return Some(Err(ArrowError::ParseError(
+                        "Error reading CSV file".to_string(),
+                    )));
+                }
+                None => break,
+            }
+        }
+
+        // return early if no data was loaded
+        if rows.len() == 0 {
+            return None;
+        }
+
+        let projection: Vec<usize> = match self.projection {
+            Some(ref v) => v.clone(),
+            None => self
+                .schema
+                .fields()
+                .iter()
+                .enumerate()
+                .map(|(i, _)| i)
+                .collect(),
+        };
+
+        let arrays: Result<Vec<ArrayRef>, ArrowError> = projection
+            .iter()
+            .map(|i| {
+                let field = self.schema.field(*i);
+
+                match field.data_type() {
+                    &DataType::Boolean => build_primitive_array!(rows, i, bool),
+                    &DataType::Int8 => build_primitive_array!(rows, i, i8),
+                    &DataType::Int16 => build_primitive_array!(rows, i, i16),
+                    &DataType::Int32 => build_primitive_array!(rows, i, i32),
+                    &DataType::Int64 => build_primitive_array!(rows, i, i64),
+                    &DataType::UInt8 => build_primitive_array!(rows, i, u8),
+                    &DataType::UInt16 => build_primitive_array!(rows, i, u16),
+                    &DataType::UInt32 => build_primitive_array!(rows, i, u32),
+                    &DataType::UInt64 => build_primitive_array!(rows, i, u64),
+                    &DataType::Float16 => build_primitive_array!(rows, i, f32),
+                    &DataType::Float32 => build_primitive_array!(rows, i, f32),
+                    &DataType::Float64 => build_primitive_array!(rows, i, f64),
+                    &DataType::Utf8 => {
+                        let mut values_builder: PrimitiveArrayBuilder<u8> =
+                            PrimitiveArrayBuilder::<u8>::new(rows.len() as i64);
+                        let mut list_builder = ListArrayBuilder::new(values_builder);
+                        for row_index in 0..rows.len() {
+                            match rows[row_index].get(*i) {
+                                Some(s) => {
+                                    list_builder.values().push_slice(s.as_bytes()).unwrap();
+                                    list_builder.append(true).unwrap();
+                                }
+                                _ => {
+                                    list_builder.append(false).unwrap();
+                                }
```

Comment:
We can update this to use `BinaryArrayBuilder` once merged.

Reply:
I pulled in 3787 and changed this to create `BinaryArray`